{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0521a6d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1332b65d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SampleID</th>\n",
       "      <th>Haplogroup</th>\n",
       "      <th>Rank</th>\n",
       "      <th>Quality</th>\n",
       "      <th>ID</th>\n",
       "      <th>Sequence_method</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KU682980</td>\n",
       "      <td>I2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0092</td>\n",
       "      <td>KU682980</td>\n",
       "      <td>WHOLE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KU683240</td>\n",
       "      <td>I2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0092</td>\n",
       "      <td>KU683240</td>\n",
       "      <td>WHOLE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KU683033</td>\n",
       "      <td>I2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0092</td>\n",
       "      <td>KU683033</td>\n",
       "      <td>WHOLE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KU683502</td>\n",
       "      <td>I2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0092</td>\n",
       "      <td>KU683502</td>\n",
       "      <td>WHOLE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EF556173</td>\n",
       "      <td>L5a1a</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0080</td>\n",
       "      <td>EF556173</td>\n",
       "      <td>WHOLE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39663</th>\n",
       "      <td>KX456745</td>\n",
       "      <td>N21+195</td>\n",
       "      <td>1</td>\n",
       "      <td>0.9500</td>\n",
       "      <td>KX456745</td>\n",
       "      <td>WHOLE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39664</th>\n",
       "      <td>SD244</td>\n",
       "      <td>N9a2'4'5'11</td>\n",
       "      <td>1</td>\n",
       "      <td>0.9500</td>\n",
       "      <td>SD244</td>\n",
       "      <td>WHOLE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39665</th>\n",
       "      <td>MN849792</td>\n",
       "      <td>Q1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.9500</td>\n",
       "      <td>MN849792</td>\n",
       "      <td>WHOLE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39667</th>\n",
       "      <td>111-1916-8482</td>\n",
       "      <td>U3a</td>\n",
       "      <td>1</td>\n",
       "      <td>0.9500</td>\n",
       "      <td>111-1916-8482</td>\n",
       "      <td>WHOLE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39668</th>\n",
       "      <td>MF523048</td>\n",
       "      <td>W3b</td>\n",
       "      <td>1</td>\n",
       "      <td>0.9500</td>\n",
       "      <td>MF523048</td>\n",
       "      <td>WHOLE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34994 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            SampleID   Haplogroup  Rank  Quality             ID  \\\n",
       "0           KU682980           I2     1   1.0092       KU682980   \n",
       "1           KU683240           I2     1   1.0092       KU683240   \n",
       "2           KU683033           I2     1   1.0092       KU683033   \n",
       "3           KU683502           I2     1   1.0092       KU683502   \n",
       "4           EF556173        L5a1a     1   1.0080       EF556173   \n",
       "...              ...          ...   ...      ...            ...   \n",
       "39663       KX456745      N21+195     1   0.9500       KX456745   \n",
       "39664          SD244  N9a2'4'5'11     1   0.9500          SD244   \n",
       "39665       MN849792           Q1     1   0.9500       MN849792   \n",
       "39667  111-1916-8482          U3a     1   0.9500  111-1916-8482   \n",
       "39668       MF523048          W3b     1   0.9500       MF523048   \n",
       "\n",
       "      Sequence_method  \n",
       "0               WHOLE  \n",
       "1               WHOLE  \n",
       "2               WHOLE  \n",
       "3               WHOLE  \n",
       "4               WHOLE  \n",
       "...               ...  \n",
       "39663           WHOLE  \n",
       "39664           WHOLE  \n",
       "39665           WHOLE  \n",
       "39667           WHOLE  \n",
       "39668           WHOLE  \n",
       "\n",
       "[34994 rows x 6 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_基础 = pd.read_csv('../data/现代线粒体DNA汇总.csv').loc[:,['ID','Sequence_method']]\n",
    "df_基础 = df_基础.loc[df_基础['Sequence_method']=='WHOLE']\n",
    "df_质量控制 = pd.read_csv('../data/质量控制.csv').iloc[:,0:4]\n",
    "df_质量控制QC = df_质量控制.loc[df_质量控制['Quality']>=0.95]\n",
    "df_质量控制QC = df_质量控制QC.merge(df_基础,left_on='SampleID',right_on='ID',how='left')\n",
    "df_质量控制QC.dropna(how='any',inplace=True)\n",
    "df_质量控制QC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "39455cc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 随机选择200个样本，要求Haplogroup不相同，且Haplogroup列要有至少1个以L开头的样本\n",
    "# 1. 每个Haplogroup保留一个样本\n",
    "df_unique = df_质量控制QC.drop_duplicates(subset=['Haplogroup'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "25a5ba73",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2. 找出以L开头的Haplogroup样本\n",
    "df_L = df_unique[df_unique['Haplogroup'].str.startswith('L')]\n",
    "\n",
    "if df_L.empty:\n",
    "    raise ValueError(\"数据中没有以L开头的Haplogroup样本，无法满足要求！\")\n",
    "\n",
    "# 3. 从以L开头的样本中随机选取1个\n",
    "sample_L = df_L.sample(1, random_state=42)\n",
    "\n",
    "# 4. 剩余样本中去掉已经选过的Haplogroup\n",
    "df_remaining = df_unique[~df_unique['Haplogroup'].isin(sample_L['Haplogroup'])]\n",
    "\n",
    "# 5. 从剩余的样本中随机选择(200-1)个\n",
    "if len(df_remaining) < 199:\n",
    "    raise ValueError(\"可选的独立Haplogroup数量不足，无法抽取199个样本！\")\n",
    "\n",
    "sample_rest = df_remaining.sample(199, random_state=42)\n",
    "\n",
    "# 6. 合并最终的200个样本\n",
    "df_final = pd.concat([sample_L, sample_rest]).reset_index(drop=True)\n",
    "\n",
    "# 结果\n",
    "df_final.sort_values(by=['Haplogroup']).loc[:,['ID','Haplogroup']].to_csv('../data/Sample_Hap_list.txt',sep='\\t',index=False,header=None)\n",
    "df_final.sort_values(by=['Haplogroup']).loc[:,['ID']].to_csv('../data/Sample_list.txt',sep='\\t',index=False,header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "80c5aa8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "!bcftools view -S /mnt/f/0_现代DNA处理流程/script/2_ML_MRCA/data/Sample_list.txt \\\n",
    "    -Oz -o ../data/Example.vcf.gz \\\n",
    "    /mnt/f/0_现代DNA处理流程/output/Merge_VCF/merged_biallelic_AGCT_SNP_Annotated_FULL_去除hots.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "6bd6e13d",
   "metadata": {},
   "outputs": [],
   "source": [
    "!bcftools index --threads 16 ../data/Example.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "96b250e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IQ-TREE version 3.0.0 for Linux x86 64-bit built Apr  4 2025\n",
      "Developed by Bui Quang Minh, Thomas Wong, Nhan Ly-Trong, Huaiyan Ren\n",
      "Contributed by Lam-Tung Nguyen, Dominik Schrempf, Chris Bielow,\n",
      "Olga Chernomor, Michael Woodhams, Diep Thi Hoang, Heiko Schmidt\n",
      "\n",
      "Host:    GiantWhiteLin (AVX2, FMA3, 31 GB RAM)\n",
      "Command: /mnt/e/Scientifc_software/iqtree-3.0.0-Linux-intel/bin/iqtree3 -redo -m HKY -s ../data/Example_rename.fasta -T 16 -o KC346074_L0d1b1a -pre ../data//Example\n",
      "Seed:    242124 (Using SPRNG - Scalable Parallel Random Number Generator)\n",
      "Time:    Mon Apr 28 22:07:00 2025\n",
      "Kernel:  AVX+FMA - 16 threads (24 CPU cores detected)\n",
      "\n",
      "Reading alignment file ../data/Example_rename.fasta ... Fasta format detected\n",
      "Reading fasta file: done in 0.026032 secs using 54.95% CPU\n",
      "Alignment most likely contains DNA/RNA sequences\n",
      "Constructing alignment: done in 0.0105229 secs using 1795% CPU\n",
      "Alignment has 200 sequences with 9419 columns, 700 distinct patterns\n",
      "381 parsimony-informative, 534 singleton sites, 8504 constant sites\n",
      "WARNING: Some sequence names are changed as follows:\n",
      "111-1907-4981_A+152+16362+200 -> 111-1907-4981_A_152_16362_200\n",
      "KU683215_H16+152 -> KU683215_H16_152\n",
      "JQ324574_H3+16189 -> JQ324574_H3_16189\n",
      "JQ703113_HV0+195 -> JQ703113_HV0_195\n",
      "FJ348153_J1c+16261 -> FJ348153_J1c_16261\n",
      "MF523173_K1b1a1+199 -> MF523173_K1b1a1_199\n",
      "EU443506_M2a1a+207 -> EU443506_M2a1a_207\n",
      "FJ383542_M3c+152 -> FJ383542_M3c_152\n",
      "JQ704045_U5a1b+16362 -> JQ704045_U5a1b_16362\n",
      "\n",
      "                               Gap/Ambiguity  Composition  p-value\n",
      "Analyzing sequences: done in 4.601e-05 secs using 7.187e+04% CPU\n",
      "   1  111-1907-4981_A_152_16362_200    0.00%    passed     99.88%\n",
      "   2  JQ703852_B2a4                    0.00%    passed     99.96%\n",
      "   3  EU095215_B2h                     0.00%    passed     99.98%\n",
      "   4  DQ372871_B4a1a                   0.00%    passed     99.99%\n",
      "   5  JX900769_B4a1a1a1a               0.00%    passed     99.89%\n",
      "   6  KJ154487_B4a1a1a1c               0.00%    passed     99.93%\n",
      "   7  MF055755_B4a1a1b                 0.00%    passed     99.99%\n",
      "   8  JX900738_B4a1a1o                 0.00%    passed     99.98%\n",
      "   9  HQ873543_B4a1a3a1                0.00%    passed     99.98%\n",
      "  10  AP008595_B4a1b                   0.00%    passed     99.98%\n",
      "  11  AY519492_B4a1c2                  0.00%    passed     99.99%\n",
      "  12  HG01866_B4b1c1                   0.00%    passed    100.00%\n",
      "  13  KC521454_B4c1a                   0.00%    passed     99.99%\n",
      "  14  AP008482_B4c1a1a                 0.00%    passed    100.00%\n",
      "  15  KC994085_B4c1b2a2a               0.00%    passed     99.99%\n",
      "  16  AY950286_B5a1a1                  0.00%    passed     99.97%\n",
      "  17  DQ372869_B5a2a2a1                0.00%    passed    100.00%\n",
      "  18  KC994058_B5b1c1a                 0.00%    passed     99.77%\n",
      "  19  AY519496_C1a                     0.00%    passed     99.94%\n",
      "  20  JX413039_C1b13a                  0.00%    passed     99.99%\n",
      "  21  DQ282447_C1b2                    0.00%    passed     99.99%\n",
      "  22  JQ705676_C1c1b                   0.00%    passed     99.99%\n",
      "  23  HM036567_C4a1a5                  0.00%    passed     99.84%\n",
      "  24  FJ383631_C4a2b1                  0.00%    passed     99.63%\n",
      "  25  EU482313_C4b7                    0.00%    passed     99.98%\n",
      "  26  111-1923-7416_D4a3b1             0.00%    passed     99.95%\n",
      "  27  AP008287_D4b2a2a1                0.00%    passed     99.98%\n",
      "  28  AP013251_D4b2b1c                 0.00%    passed     99.99%\n",
      "  29  111-1907-1393_D4b2b7             0.00%    passed     99.99%\n",
      "  30  AP008628_D4d                     0.00%    passed     99.99%\n",
      "  31  HQ012262_D4e1c                   0.00%    passed     99.90%\n",
      "  32  111-1937-0434_D4e5               0.00%    passed     99.98%\n",
      "  33  111-1914-6930_D4g2b              0.00%    passed    100.00%\n",
      "  34  FJ168712_D4h3b                   0.00%    passed     99.89%\n",
      "  35  FJ383234_D4q1                    0.00%    passed    100.00%\n",
      "  36  111-1921-0792_D5b3               0.00%    passed     99.99%\n",
      "  37  111-1910-8495_F                  0.00%    passed     99.99%\n",
      "  38  KX456435_F1a1a1                  0.00%    passed     99.79%\n",
      "  39  111-1919-0674_F1b1               0.00%    passed     99.90%\n",
      "  40  KF849909_F2i                     0.00%    passed     99.90%\n",
      "  41  111-1945-6527_F3a1               0.00%    passed     99.99%\n",
      "  42  111-1903-2646_G1a1a              0.00%    passed     99.99%\n",
      "  43  AP008909_G1a1a3                  0.00%    passed    100.00%\n",
      "  44  111-1926-1381_G2a3a              0.00%    passed     99.91%\n",
      "  45  111-1936-0787_G3a                0.00%    passed    100.00%\n",
      "  46  EU073971_H10a1a1                 0.00%    passed     99.99%\n",
      "  47  ERM088_H10e1                     0.00%    passed     99.99%\n",
      "  48  JQ703386_H13a1a1b                0.00%    passed    100.00%\n",
      "  49  JQ704313_H13b2                   0.00%    passed    100.00%\n",
      "  50  KU683215_H16_152                 0.00%    passed    100.00%\n",
      "  51  HM852833_H19                     0.00%    passed     99.99%\n",
      "  52  AY339422_H1a2                    0.00%    passed    100.00%\n",
      "  53  JQ702057_H1aa1                   0.00%    passed    100.00%\n",
      "  54  JQ701969_H1ai1                   0.00%    passed    100.00%\n",
      "  55  JQ702335_H1an2                   0.00%    passed     99.99%\n",
      "  56  JQ704370_H1as2                   0.00%    passed    100.00%\n",
      "  57  JQ705362_H1av                    0.00%    passed    100.00%\n",
      "  58  JQ898578_H1bz                    0.00%    passed    100.00%\n",
      "  59  JQ705712_H1c                     0.00%    passed    100.00%\n",
      "  60  JX153093_H1ch                    0.00%    passed    100.00%\n",
      "  61  JX153389_H1ck                    0.00%    passed    100.00%\n",
      "  62  JX297199_H1e1a6                  0.00%    passed    100.00%\n",
      "  63  JX153958_H1e1a7                  0.00%    passed    100.00%\n",
      "  64  JX153041_H1e5b                   0.00%    passed    100.00%\n",
      "  65  JQ703312_H1n5                    0.00%    passed    100.00%\n",
      "  66  EF556184_H1p                     0.00%    passed    100.00%\n",
      "  67  FJ460532_H1v                     0.00%    passed    100.00%\n",
      "  68  JQ704562_H1z1                    0.00%    passed    100.00%\n",
      "  69  EU007878_H27a                    0.00%    passed    100.00%\n",
      "  70  JQ704644_H27f                    0.00%    passed    100.00%\n",
      "  71  JQ701848_H2a1a2                  0.00%    passed     99.99%\n",
      "  72  JQ704847_H2a1e1b                 0.00%    passed    100.00%\n",
      "  73  EF418606_H2a2b1a1                0.00%    passed    100.00%\n",
      "  74  JQ324574_H3_16189                0.00%    passed    100.00%\n",
      "  75  JQ703157_H35                     0.00%    passed    100.00%\n",
      "  76  JQ702256_H35a                    0.00%    passed     99.98%\n",
      "  77  JQ703140_H3ap                    0.00%    passed     99.99%\n",
      "  78  JQ704192_H3as                    0.00%    passed    100.00%\n",
      "  79  JQ705041_H3h5                    0.00%    passed     99.99%\n",
      "  80  KF148401_H49                     0.00%    passed     99.97%\n",
      "  81  JQ702985_H49a1                   0.00%    passed     99.99%\n",
      "  82  JQ702960_H4c1                    0.00%    passed    100.00%\n",
      "  83  AY495174_H5                      0.00%    passed    100.00%\n",
      "  84  KC257389_H56d                    0.00%    passed    100.00%\n",
      "  85  EU677750_H5e1a                   0.00%    passed     99.96%\n",
      "  86  GQ983082_H5e1b                   0.00%    passed    100.00%\n",
      "  87  JQ703547_H5s                     0.00%    passed    100.00%\n",
      "  88  JX153784_H6a1b2e                 0.00%    passed    100.00%\n",
      "  89  JQ324534_H87                     0.00%    passed     99.98%\n",
      "  90  JQ703113_HV0_195                 0.00%    passed    100.00%\n",
      "  91  GU122990_HV0a1                   0.00%    passed    100.00%\n",
      "  92  AY713976_HV12b                   0.00%    passed    100.00%\n",
      "  93  JQ703367_HV9                     0.00%    passed    100.00%\n",
      "  94  KF146244_I1c1                    0.00%    passed     99.79%\n",
      "  95  JQ245724_I5                      0.00%    passed     99.99%\n",
      "  96  DQ112955_I6a                     0.00%    passed    100.00%\n",
      "  97  AY714033_J1b1b3                  0.00%    passed     99.99%\n",
      "  98  JQ797777_J1b5a1                  0.00%    passed     99.97%\n",
      "  99  HM852835_J1b9                    0.00%    passed    100.00%\n",
      " 100  FJ348153_J1c_16261               0.00%    passed     99.99%\n",
      " 101  JQ704751_J1c10                   0.00%    passed    100.00%\n",
      " 102  JQ704432_J1c2f                   0.00%    passed    100.00%\n",
      " 103  JQ705996_J1c2h                   0.00%    passed    100.00%\n",
      " 104  JQ705054_J1c2t                   0.00%    passed    100.00%\n",
      " 105  KY409278_J1c3e2                  0.00%    passed     99.99%\n",
      " 106  HGDP01396_J1c3k                  0.00%    passed     99.99%\n",
      " 107  JQ797836_J1c3m                   0.00%    passed     99.98%\n",
      " 108  MF523006_J1d2                    0.00%    passed    100.00%\n",
      " 109  JQ797920_J2a2a1a                 0.00%    passed     99.90%\n",
      " 110  HGDP00623_J2a2a1a1               0.00%    passed     99.94%\n",
      " 111  JQ797933_J2a2c1                  0.00%    passed     99.85%\n",
      " 112  FJ445409_J2b1d                   0.00%    passed     99.99%\n",
      " 113  FJ213765_J2b2                    0.00%    passed    100.00%\n",
      " 114  JQ706038_K1a                     0.00%    passed     99.92%\n",
      " 115  JQ706039_K1a19a                  0.00%    passed     99.85%\n",
      " 116  HGDP01361_K1a1b1                 0.00%    passed     99.85%\n",
      " 117  KC911309_K1a31                   0.00%    passed     99.87%\n",
      " 118  JQ705957_K1a3a2                  0.00%    passed     99.87%\n",
      " 119  MF523173_K1b1a1_199              0.00%    passed     99.92%\n",
      " 120  JQ702168_K1b1a1a                 0.00%    passed     99.99%\n",
      " 121  DQ301815_K2a1a                   0.00%    passed     99.92%\n",
      " 122  JN409346_K2b1a1a                 0.00%    passed     99.97%\n",
      " 123  KC346074_L0d1b1a                 0.00%    passed     98.81%\n",
      " 124  KC345900_L0d1b2b1a               0.00%    passed     98.78%\n",
      " 125  KC345781_L0d1b2b2                0.00%    passed     98.53%\n",
      " 126  EU092852_L1b1a10b                0.00%    passed     99.96%\n",
      " 127  EU092737_L1b1a9                  0.00%    passed     99.98%\n",
      " 128  EU273490_L1c1a1a1a               0.00%    passed     99.84%\n",
      " 129  DQ304933_L2a1a1                  0.00%    passed    100.00%\n",
      " 130  JQ412577_L2a1c1a2                0.00%    passed     99.97%\n",
      " 131  JX303838_L2a1d2                  0.00%    passed     99.96%\n",
      " 132  EU092671_L2a1p                   0.00%    passed    100.00%\n",
      " 133  EU092813_L2c1a                   0.00%    passed     98.96%\n",
      " 134  JN214459_L3d1b1a                 0.00%    passed    100.00%\n",
      " 135  EU092830_L3d1c1                  0.00%    passed     99.87%\n",
      " 136  HM771135_L3e1f1                  0.00%    passed    100.00%\n",
      " 137  KM986577_L3f2a1a                 0.00%    passed     99.99%\n",
      " 138  KM986620_L3x1a1                  0.00%    passed    100.00%\n",
      " 139  EF184627_L4b2a2                  0.00%    passed     99.95%\n",
      " 140  PP654541_M17c1a1a                0.00%    passed     99.93%\n",
      " 141  EF060333_M1a1e1                  0.00%    passed     99.96%\n",
      " 142  MF055751_M23                     0.00%    passed     99.98%\n",
      " 143  DQ372879_M28a1                   0.00%    passed     99.98%\n",
      " 144  KJ154481_M28a6a                  0.00%    passed     99.97%\n",
      " 145  EU443506_M2a1a_207               0.00%    passed     99.93%\n",
      " 146  FJ383542_M3c_152                 0.00%    passed     99.99%\n",
      " 147  DQ112875_M5a1a                   0.00%    passed     99.83%\n",
      " 148  AP013148_M7a1b2                  0.00%    passed    100.00%\n",
      " 149  111-1929-8585_M7b                0.00%    passed    100.00%\n",
      " 150  111-1906-2575_M7b1a1a            0.00%    passed     99.99%\n",
      " 151  111-1917-2649_M7b1a1a1           0.00%    passed     99.93%\n",
      " 152  111-1920-7815_M7b1a1b            0.00%    passed     99.99%\n",
      " 153  111-1922-8506_M7b1a1h            0.00%    passed     99.98%\n",
      " 154  111-1922-3495_M7c1               0.00%    passed     99.99%\n",
      " 155  111-1908-9107_M9a1a1c1a          0.00%    passed     99.89%\n",
      " 156  FJ383311_M9a1b1a                 0.00%    passed     99.99%\n",
      " 157  JQ701924_N1a1a2                  0.00%    passed    100.00%\n",
      " 158  DQ301794_N1b1b1                  0.00%    passed     99.72%\n",
      " 159  111-1122-3363_N9a1               0.00%    passed     99.93%\n",
      " 160  111-1936-3290_N9a2d              0.00%    passed     99.99%\n",
      " 161  AP010722_N9b1                    0.00%    passed     99.95%\n",
      " 162  EF556170_R0a2a                   0.00%    passed    100.00%\n",
      " 163  HM185224_R0a3a                   0.00%    passed     99.99%\n",
      " 164  JF944823_T2a2a                   0.00%    passed    100.00%\n",
      " 165  AY495287_T2b5                    0.00%    passed     99.96%\n",
      " 166  JQ702795_T2c1c2                  0.00%    passed    100.00%\n",
      " 167  JQ798108_T2d2                    0.00%    passed     99.99%\n",
      " 168  AY714029_T2e2                    0.00%    passed     99.98%\n",
      " 169  JQ798110_T2e7                    0.00%    passed     99.99%\n",
      " 170  JX153995_T2f7                    0.00%    passed    100.00%\n",
      " 171  JQ705601_U1a1a3                  0.00%    passed     99.98%\n",
      " 172  KC911365_U1a1c1c                 0.00%    passed    100.00%\n",
      " 173  HM852858_U1b                     0.00%    passed     99.97%\n",
      " 174  FJ348194_U2e1a1c                 0.00%    passed     99.53%\n",
      " 175  111-1937-6535_U4a                0.00%    passed     99.97%\n",
      " 176  JQ702549_U4a2h1                  0.00%    passed     99.99%\n",
      " 177  EU545423_U4a3a                   0.00%    passed     99.99%\n",
      " 178  FJ147313_U4b1a4                  0.00%    passed     99.96%\n",
      " 179  JQ704045_U5a1b_16362             0.00%    passed     99.97%\n",
      " 180  JQ705245_U5a1b1a1                0.00%    passed     99.97%\n",
      " 181  KC569552_U5a1i1                  0.00%    passed     99.95%\n",
      " 182  GU296626_U5a2a1b                 0.00%    passed    100.00%\n",
      " 183  JQ704044_U5a2b1b                 0.00%    passed    100.00%\n",
      " 184  GU296587_U5a2b2a1                0.00%    passed     99.99%\n",
      " 185  JX152982_U5b1b2b                 0.00%    passed    100.00%\n",
      " 186  JX153147_U5b2a3a                 0.00%    passed     99.98%\n",
      " 187  JQ705947_U5b2c2a                 0.00%    passed     99.93%\n",
      " 188  GQ129143_U5b3g                   0.00%    passed     99.99%\n",
      " 189  KC152549_U6a3a2                  0.00%    passed     99.99%\n",
      " 190  KC152585_U6a3f2                  0.00%    passed     99.98%\n",
      " 191  EF064334_U6a6b2                  0.00%    passed     99.95%\n",
      " 192  AY275528_U6b1a1                  0.00%    passed     99.99%\n",
      " 193  HM775494_U6c2                    0.00%    passed     99.82%\n",
      " 194  AY339451_V5                      0.00%    passed     99.98%\n",
      " 195  MN595741_W3a1a                   0.00%    passed     99.96%\n",
      " 196  EU439939_X2a1                    0.00%    passed     99.92%\n",
      " 197  JQ704754_X2b7                    0.00%    passed     99.85%\n",
      " 198  JX508853_X2p1                    0.00%    passed     99.90%\n",
      " 199  111-1926-6935_Y1                 0.00%    passed    100.00%\n",
      " 200  GQ119016_Y2a                     0.00%    passed     99.99%\n",
      "****  TOTAL                            0.00%  0 sequences failed composition chi2 test (p-value<5%; df=3)\n",
      "NOTE: JQ703157_H35 is identical to KU683215_H16_152 but kept for subsequent analysis\n",
      "NOTE: FJ460532_H1v is identical to EF556184_H1p but kept for subsequent analysis\n",
      "NOTE: JQ704192_H3as is identical to JQ324574_H3_16189 but kept for subsequent analysis\n",
      "Checking for duplicate sequences: done in 0.00363335 secs using 614.1% CPU\n",
      "\n",
      "Create initial parsimony tree by phylogenetic likelihood library (PLL)... 0.019 seconds\n",
      "\n",
      "NOTE: 4 MB RAM (0 GB) is required!\n",
      "WARNING: Number of threads seems too high for short alignments. Use -T AUTO to determine best number of threads.\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "1. Initial log-likelihood: -24596.274\n",
      "2. Current log-likelihood: -23441.225\n",
      "Optimal log-likelihood: -23441.232\n",
      "Rate parameters:  A-C: 1.00000  A-G: 100.00000  A-T: 1.00000  C-G: 1.00000  C-T: 100.00000  G-T: 1.00000\n",
      "Warning! Some parameters hit the boundaries\n",
      "Base frequencies:  A: 0.302  C: 0.306  G: 0.185  T: 0.206\n",
      "Parameters optimization took 2 rounds (0.039 sec)\n",
      "\n",
      "WARNING: Estimated model parameters are at boundary that can cause numerical instability!\n",
      "\n",
      "Wrote distance file to... \n",
      "Computing ML distances based on estimated model parameters...\n",
      "Calculating distance matrix: done in 0.00568107 secs using 3196% CPU\n",
      "Computing ML distances took 0.006134 sec (of wall-clock time) 0.181779 sec (of CPU time)\n",
      "Setting up auxiliary I and S matrices: done in 0.000617255 secs using 105.3% CPU\n",
      "Constructing RapidNJ tree: done in 0.00155125 secs using 127.1% CPU\n",
      "Computing RapidNJ tree took 0.002406 sec (of wall-clock time) 0.002768 sec (of CPU time)\n",
      "Log-likelihood of RapidNJ tree: -23448.061\n",
      "--------------------------------------------------------------------\n",
      "|             INITIALIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Generating 98 parsimony trees... 1.696 second\n",
      "Computing log-likelihood of 98 initial trees ... 0.589 seconds\n",
      "Current best score: -23439.993\n",
      "\n",
      "Do NNI search on 20 best initial trees\n",
      "Optimizing NNI: done in 0.0730527 secs using 1529% CPU\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "\n",
      "WARNING: Estimated model parameters are at boundary that can cause numerical instability!\n",
      "\n",
      "BETTER TREE FOUND at iteration 1: -23435.886\n",
      "Optimizing NNI: done in 0.072794 secs using 1507% CPU\n",
      "Optimizing NNI: done in 0.0875387 secs using 1471% CPU\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "\n",
      "WARNING: Estimated model parameters are at boundary that can cause numerical instability!\n",
      "\n",
      "BETTER TREE FOUND at iteration 3: -23435.867\n",
      "Optimizing NNI: done in 0.0775982 secs using 1634% CPU\n",
      "Optimizing NNI: done in 0.100845 secs using 1718% CPU\n",
      "Optimizing NNI: done in 0.107034 secs using 1501% CPU\n",
      "Optimizing NNI: done in 0.105282 secs using 1515% CPU\n",
      "Optimizing NNI: done in 0.0752035 secs using 1682% CPU\n",
      "Optimizing NNI: done in 0.0862291 secs using 1666% CPU\n",
      "Optimizing NNI: done in 0.0838012 secs using 1532% CPU\n",
      "Iteration 10 / LogL: -23440.784 / Time: 0h:0m:3s\n",
      "Optimizing NNI: done in 0.0807578 secs using 1586% CPU\n",
      "Optimizing NNI: done in 0.100455 secs using 1593% CPU\n",
      "Optimizing NNI: done in 0.0834544 secs using 1526% CPU\n",
      "Optimizing NNI: done in 0.0847788 secs using 1692% CPU\n",
      "Optimizing NNI: done in 0.105622 secs using 1520% CPU\n",
      "Optimizing NNI: done in 0.11167 secs using 1569% CPU\n",
      "Optimizing NNI: done in 0.0607704 secs using 1565% CPU\n",
      "Optimizing NNI: done in 0.0754347 secs using 1691% CPU\n",
      "Optimizing NNI: done in 0.0816833 secs using 1569% CPU\n",
      "Optimizing NNI: done in 0.113986 secs using 1679% CPU\n",
      "Iteration 20 / LogL: -23435.943 / Time: 0h:0m:4s\n",
      "Finish initializing candidate tree set (20)\n",
      "Current best tree score: -23435.867 / CPU time: 4.118\n",
      "Number of iterations: 20\n",
      "--------------------------------------------------------------------\n",
      "|               OPTIMIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Optimizing NNI: done in 0.156599 secs using 1510% CPU\n",
      "Optimizing NNI: done in 0.140074 secs using 1592% CPU\n",
      "Optimizing NNI: done in 0.144555 secs using 1637% CPU\n",
      "Optimizing NNI: done in 0.103439 secs using 1650% CPU\n",
      "Optimizing NNI: done in 0.147582 secs using 1618% CPU\n",
      "Optimizing NNI: done in 0.163455 secs using 1644% CPU\n",
      "Optimizing NNI: done in 0.159721 secs using 1570% CPU\n",
      "Optimizing NNI: done in 0.15009 secs using 1595% CPU\n",
      "Optimizing NNI: done in 0.134744 secs using 1659% CPU\n",
      "Optimizing NNI: done in 0.140848 secs using 1563% CPU\n",
      "Iteration 30 / LogL: -23436.365 / Time: 0h:0m:5s (0h:0m:14s left)\n",
      "Optimizing NNI: done in 0.172933 secs using 1562% CPU\n",
      "Optimizing NNI: done in 0.145112 secs using 1531% CPU\n",
      "Optimizing NNI: done in 0.174736 secs using 1548% CPU\n",
      "Optimizing NNI: done in 0.159111 secs using 1593% CPU\n",
      "Optimizing NNI: done in 0.151675 secs using 1577% CPU\n",
      "Optimizing NNI: done in 0.131435 secs using 1564% CPU\n",
      "Optimizing NNI: done in 0.135297 secs using 1579% CPU\n",
      "Optimizing NNI: done in 0.139753 secs using 1732% CPU\n",
      "Optimizing NNI: done in 0.108603 secs using 1606% CPU\n",
      "Optimizing NNI: done in 0.130672 secs using 1730% CPU\n",
      "Iteration 40 / LogL: -23439.970 / Time: 0h:0m:7s (0h:0m:11s left)\n",
      "Optimizing NNI: done in 0.134908 secs using 1760% CPU\n",
      "Optimizing NNI: done in 0.147606 secs using 1744% CPU\n",
      "Optimizing NNI: done in 0.127631 secs using 1760% CPU\n",
      "Optimizing NNI: done in 0.122467 secs using 1703% CPU\n",
      "Optimizing NNI: done in 0.144321 secs using 1693% CPU\n",
      "Optimizing NNI: done in 0.125342 secs using 1666% CPU\n",
      "Optimizing NNI: done in 0.130041 secs using 1741% CPU\n",
      "Optimizing NNI: done in 0.147756 secs using 1733% CPU\n",
      "Optimizing NNI: done in 0.149741 secs using 1632% CPU\n",
      "Optimizing NNI: done in 0.147665 secs using 1752% CPU\n",
      "Iteration 50 / LogL: -23436.171 / Time: 0h:0m:8s (0h:0m:9s left)\n",
      "Optimizing NNI: done in 0.11967 secs using 1717% CPU\n",
      "Optimizing NNI: done in 0.119058 secs using 1758% CPU\n",
      "Optimizing NNI: done in 0.109102 secs using 1723% CPU\n",
      "Optimizing NNI: done in 0.152903 secs using 1814% CPU\n",
      "Optimizing NNI: done in 0.121414 secs using 1709% CPU\n",
      "Optimizing NNI: done in 0.132607 secs using 1705% CPU\n",
      "Optimizing NNI: done in 0.124874 secs using 1804% CPU\n",
      "Optimizing NNI: done in 0.127572 secs using 1745% CPU\n",
      "Optimizing NNI: done in 0.11317 secs using 1693% CPU\n",
      "Optimizing NNI: done in 0.140776 secs using 1721% CPU\n",
      "Iteration 60 / LogL: -23436.211 / Time: 0h:0m:10s (0h:0m:7s left)\n",
      "Optimizing NNI: done in 0.134489 secs using 1691% CPU\n",
      "Optimizing NNI: done in 0.127767 secs using 1762% CPU\n",
      "Optimizing NNI: done in 0.133118 secs using 1699% CPU\n",
      "Optimizing NNI: done in 0.120994 secs using 1719% CPU\n",
      "Optimizing NNI: done in 0.117553 secs using 1640% CPU\n",
      "Optimizing NNI: done in 0.151557 secs using 1726% CPU\n",
      "Optimizing NNI: done in 0.145272 secs using 1798% CPU\n",
      "Optimizing NNI: done in 0.140115 secs using 1744% CPU\n",
      "Optimizing NNI: done in 0.128363 secs using 1752% CPU\n",
      "Optimizing NNI: done in 0.118394 secs using 1751% CPU\n",
      "Iteration 70 / LogL: -23444.847 / Time: 0h:0m:11s (0h:0m:5s left)\n",
      "Optimizing NNI: done in 0.130866 secs using 1827% CPU\n",
      "Optimizing NNI: done in 0.139151 secs using 1748% CPU\n",
      "Optimizing NNI: done in 0.134059 secs using 1774% CPU\n",
      "Optimizing NNI: done in 0.131303 secs using 1722% CPU\n",
      "Optimizing NNI: done in 0.120325 secs using 1722% CPU\n",
      "Optimizing NNI: done in 0.0990568 secs using 1595% CPU\n",
      "Optimizing NNI: done in 0.129077 secs using 1734% CPU\n",
      "Optimizing NNI: done in 0.118121 secs using 1741% CPU\n",
      "Optimizing NNI: done in 0.139257 secs using 1500% CPU\n",
      "Optimizing NNI: done in 0.140068 secs using 1585% CPU\n",
      "Iteration 80 / LogL: -23439.819 / Time: 0h:0m:12s (0h:0m:3s left)\n",
      "Optimizing NNI: done in 0.158415 secs using 1615% CPU\n",
      "Optimizing NNI: done in 0.146166 secs using 1525% CPU\n",
      "Optimizing NNI: done in 0.142009 secs using 1575% CPU\n",
      "Optimizing NNI: done in 0.1471 secs using 1617% CPU\n",
      "Optimizing NNI: done in 0.15031 secs using 1581% CPU\n",
      "Optimizing NNI: done in 0.13712 secs using 1515% CPU\n",
      "Optimizing NNI: done in 0.143453 secs using 1564% CPU\n",
      "Optimizing NNI: done in 0.181302 secs using 1581% CPU\n",
      "Optimizing NNI: done in 0.148916 secs using 1599% CPU\n",
      "Optimizing NNI: done in 0.137313 secs using 1603% CPU\n",
      "Iteration 90 / LogL: -23436.097 / Time: 0h:0m:14s (0h:0m:2s left)\n",
      "Optimizing NNI: done in 0.128806 secs using 1604% CPU\n",
      "Optimizing NNI: done in 0.151603 secs using 1582% CPU\n",
      "Optimizing NNI: done in 0.134521 secs using 1642% CPU\n",
      "Optimizing NNI: done in 0.128246 secs using 1613% CPU\n",
      "Optimizing NNI: done in 0.119241 secs using 1719% CPU\n",
      "Optimizing NNI: done in 0.145517 secs using 1791% CPU\n",
      "Optimizing NNI: done in 0.137046 secs using 1759% CPU\n",
      "Optimizing NNI: done in 0.121418 secs using 1828% CPU\n",
      "Optimizing NNI: done in 0.125294 secs using 1792% CPU\n",
      "Optimizing NNI: done in 0.127294 secs using 1769% CPU\n",
      "Iteration 100 / LogL: -23436.128 / Time: 0h:0m:15s (0h:0m:0s left)\n",
      "Optimizing NNI: done in 0.140075 secs using 1738% CPU\n",
      "Optimizing NNI: done in 0.129647 secs using 1742% CPU\n",
      "Optimizing NNI: done in 0.144415 secs using 1797% CPU\n",
      "Optimizing NNI: done in 0.11798 secs using 1764% CPU\n",
      "TREE SEARCH COMPLETED AFTER 104 ITERATIONS / Time: 0h:0m:16s\n",
      "\n",
      "--------------------------------------------------------------------\n",
      "|                    FINALIZING TREE SEARCH                        |\n",
      "--------------------------------------------------------------------\n",
      "Performs final model parameters optimization\n",
      "Estimate model parameters (epsilon = 0.010)\n",
      "1. Initial log-likelihood: -23435.867\n",
      "Optimal log-likelihood: -23435.867\n",
      "Rate parameters:  A-C: 1.00000  A-G: 100.00000  A-T: 1.00000  C-G: 1.00000  C-T: 100.00000  G-T: 1.00000\n",
      "Warning! Some parameters hit the boundaries\n",
      "Base frequencies:  A: 0.302  C: 0.306  G: 0.185  T: 0.206\n",
      "Parameters optimization took 1 rounds (0.016 sec)\n",
      "\n",
      "WARNING: Estimated model parameters are at boundary that can cause numerical instability!\n",
      "\n",
      "BEST SCORE FOUND : -23435.867\n",
      "Total tree length: 0.133\n",
      "\n",
      "Total number of iterations: 104\n",
      "CPU time used for tree search: 240.119 sec (0h:4m:0s)\n",
      "Wall-clock time used for tree search: 16.017 sec (0h:0m:16s)\n",
      "Total CPU time used: 242.156 sec (0h:4m:2s)\n",
      "Total wall-clock time used: 16.245 sec (0h:0m:16s)\n",
      "\n",
      "Analysis results written to: \n",
      "  IQ-TREE report:                ../data//Example.iqtree\n",
      "  Maximum-likelihood tree:       ../data//Example.treefile\n",
      "  Likelihood distances:          ../data//Example.mldist\n",
      "  Screen log file:               ../data//Example.log\n",
      "\n",
      "Date and Time: Mon Apr 28 22:07:16 2025\n"
     ]
    }
   ],
   "source": [
    "\n",
    "!/mnt/e/Scientifc_software/iqtree-3.0.0-Linux-intel/bin/iqtree3 \\\n",
    "  -redo \\\n",
    "  -m HKY \\\n",
    "  -s ../data/Example_rename.fasta \\\n",
    "  -T 16 \\\n",
    "  -o KC346074_L0d1b1a \\\n",
    "  -pre ../data//Example\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8cbfa6f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "!cat ../data/Example.treefile >> ../data/Example.boottrees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97ee760f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "已成功更新 ../data/Example.treefile ：找到 1 棵树，样本数为 197\n"
     ]
    }
   ],
   "source": [
    "from Bio import Phylo\n",
    "\n",
    "# 1. 指定树文件路径\n",
    "tree_file = '../data/Example.treefile'\n",
    "\n",
    "# 2. 读取所有 bootstrap 树\n",
    "trees = list(Phylo.parse(tree_file, 'newick'))\n",
    "num_trees = len(trees)\n",
    "\n",
    "# 3. 统计样本数（假设每棵树的叶子节点数一致，以第一棵树为准）\n",
    "num_samples = len(trees[0].get_terminals()) if num_trees > 0 else 0\n",
    "\n",
    "# 4. 读取原始文件内容（保留所有树的 Newick 文本）\n",
    "with open(tree_file, 'r') as f:\n",
    "    original_content = f.read().strip()\n",
    "\n",
    "# 5. 将统计信息和原始内容写回同一文件\n",
    "with open(tree_file, 'w') as f:\n",
    "    # 首行：树的数量 和 样本数\n",
    "    f.write(f\" {num_samples} {num_trees}\\n\")\n",
    "    # 后续行：原始的所有 bootstrap 树\n",
    "    f.write(original_content + \"\\n\")\n",
    "\n",
    "print(f\"已成功更新 {tree_file} ：找到 {num_trees} 棵树，样本数为 {num_samples}\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
